### DISCUSSION TITLE
Show & Tell: Inside the Bio-Inspired Module — STDP, circadian rhythms, sleep-wake consolidation

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#45] [Show and tell]
# Show & Tell: Inside the Bio-Inspired Module

The `bio_inspired` module has one of the richest internal structures in ASI:BUILD — 14 sub-components modelling everything from spike-timing plasticity to circadian sleep pressure. I want to walk through what it actually does and open some questions about where to take it next.

## What it is

Unlike most modules that implement specific algorithms, `bio_inspired` implements *mechanisms* — the same kinds of mechanisms that biological brains use to remain efficient, adaptable, and self-regulating over time. It sits underneath the `CognitiveCycle` as both a sensory preprocessor (Phase 1, embodied perception) and a homeostatic governor (Phase 9).

## Eight cognitive states

The system maintains a `CognitiveState` enum with 8 values: `AWAKE_ACTIVE`, `AWAKE_RESTING`, `NREM_SLEEP`, `REM_SLEEP`, `LEARNING`, `CONSOLIDATING`, `DEVELOPING`, and `ADAPTING`. State transitions are driven by a **sleep pressure accumulator** and a **circadian phase** signal:

```python
sleep_drive = sleep_pressure + 0.5 * np.sin(2 * np.pi * circadian_clock)
```

This mirrors the two-process model of sleep regulation (Borbély, 1982): Process S (adenosine buildup) + Process C (circadian rhythm). When `sleep_drive > 0.8`, the system enters NREM (80% probability) or REM (20%).

## Neuromodulation

Four neurotransmitters are tracked as continuous values:

| NT | Default | Role |
|----|---------|------|
| Dopamine | 0.3 | Reward prediction, learning gate |
| Serotonin | 0.4 | Mood, satiety, sleep preparation |
| Norepinephrine | 0.2 | Arousal, attention, fight-or-flight |
| Acetylcholine | 0.5 | Memory encoding, attention sharpening |

These modulate learning rates, plasticity thresholds, and attention weighting across other modules. High dopamine → stronger LTP. High NE → tighter attention focus. Low serotonin → emotional dysregulation during the `EmotionalRegulation` module's processing.

## STDP and BCM learning rules

`STDPLearning` implements spike-timing dependent plasticity:

- **LTP** (pre fires before post, causal): `Δw = A_plus * exp(-Δt / τ_plus)`
- **LTD** (post fires before pre, anti-causal): `Δw = -A_minus * exp(Δt / τ_minus)`

The window is 20ms, matching cortical measurements. The BCM rule adds a **sliding modification threshold** θ_M that rises when a neuron fires too often — synaptic homeostasis that prevents runaway potentiation without needing a global learning rate schedule.

## Sleep-wake and memory consolidation

The `SleepWakeCycle` uses the two-stage memory model: fast hippocampal encoding during wake, slow cortical consolidation during sleep. NREM has the strongest consolidation weight (1.0); REM handles emotional processing and pattern replay.

In a fully integrated system, the `knowledge_graph` should accept weaker writes during wake (high plasticity, noisy) and stronger filtered consolidations after sleep cycles. That bridge doesn't exist yet — it's one of the open design questions.

## Developmental stages

`DevelopmentalLearning` models four stages with `developmental_stage` as a continuous float 0→1:
- **Infant** (0–0.2): high plasticity, broad generalization
- **Child** (0.2–0.5): category formation, rapid learning
- **Adolescent** (0.5–0.8): abstract reasoning, pruning begins
- **Adult** (0.8–1.0): specialization, lower baseline plasticity

## New wiki page

Full architecture, config reference, usage example, and open questions now documented: [Bio-Inspired Architecture](https://github.com/web3guru888/asi-build/wiki/Bio-Inspired)

## Open questions for the community

1. **Sleep pressure calibration**: The current accumulator is linear (`+0.0001` per tick). Real adenosine dynamics are nonlinear with glymphatic clearance during deep sleep. How faithful should the model be for AGI purposes?

2. **STDP vs backprop**: Is there a use case where we'd want to *disable* gradient descent and rely purely on STDP? What tasks specifically benefit from temporal credit assignment over spatial (backprop)?

3. **Neuromodulator cross-coupling**: In biology, serotonin suppresses dopaminergic VTA firing. Should the `NeuromodulationSystem` model these NT-NT interactions, or does that add complexity without insight?

4. **Knowledge graph + sleep integration**: Should the KG's write path check `CognitiveState` and apply state-dependent consolidation weights? This would create a real feedback loop between bio_inspired and the knowledge graph — more faithful to biology but harder to reason about.

5. **Energy accounting**: Each spike costs 0.001 energy units. Is that abstraction meaningful, or should it be grounded in actual FLOP counts or joules for comparison against the 20W human brain?

Question 4 interests me most — it has the most direct impact on overall system behavior. If KG writes during NREM consolidation are stronger and more selective, the system effectively "sleeps to forget" irrelevant noise. That's a genuinely interesting property to test. Thoughts?
